packages/language/examples/src/atomicSwap.ts

```diff
@@ -0,0 +1,444 @@
+      const askInput: IDeposit = {
+        input_from_party: scheme.participants.buyer,
+        that_deposits: scheme.ask.asset.amount,
+        of_token: scheme.ask.asset.token,
+        into_account: scheme.participants.buyer,
+      };
+      const retractInput: IChoice = {
+        for_choice_id: {
+          choice_name: "retract",
+          choice_owner: scheme.participants.seller,
+        },
+        input_that_chooses_num: 0n,
+      };
+      return {
+        typeName: "WaitingForAnswer",
+        scheme: scheme,
+        actions: [
+          {
+            typeName: "Swap",
+            owner: "buyer",
+            input: askInput,
+          },
+          {
+            typeName: "Retract",
+            owner: "seller",
+            input: retractInput,
+          },
+        ],
+      };
+    } else {
+      // Closed (handled upstream)
+    }
+  }
+
+  if (inputHistory.length === 2) {
+    const nbDeposits = inputHistory.filter((input) =>
+      G.IDeposit.is(input)
+    ).length;
+    if (nbDeposits === 2 && now < scheme.swapConfirmation.deadline) {
+      return {
+        typeName: "WaitingForSwapConfirmation",
+        scheme: scheme,
+        action: {
+          typeName: "ConfirmSwap",
+          owner: "anybody",
+          input: "input_notify",
+        },
+      };
+    } else {
+      // Closed (handled upstream)
+    }
+  }
+
+  throw new UnexpectedSwapContractState(scheme, state);
+};
+export function mkContract(scheme: Scheme): Contract {
+  const mkOffer = (ask: Contract): Contract => {
+    const depositOffer = {
+      party: scheme.participants.seller,
+      deposits: scheme.offer.asset.amount,
+      of_token: scheme.offer.asset.token,
+      into_account: scheme.participants.seller,
+    };
+
+    return {
+      when: [{ case: depositOffer, then: ask }],
+      timeout: scheme.offer.deadline,
+      timeout_continuation: close,
+    };
+  };
+
+  const mkAsk = (confirmSwap: Contract): Contract => {
+    const asPayee = (party: Party): PayeeParty => ({ party: party });
+    const depositAsk = {
+      party: scheme.participants.buyer,
+      deposits: scheme.ask.asset.amount,
+      of_token: scheme.ask.asset.token,
+      into_account: scheme.participants.buyer,
+    };
+    const chooseToRetract = {
+      choose_between: [{ from: 0n, to: 0n }],
+      for_choice: {
+        choice_name: "retract",
+        choice_owner: scheme.participants.seller,
+      },
+    };
+    return {
+      when: [
+        {
+          case: depositAsk,
+          then: {
+            pay: scheme.offer.asset.amount,
+            token: scheme.offer.asset.token,
+            from_account: scheme.participants.seller,
+            to: asPayee(scheme.participants.buyer),
+            then: {
+              pay: scheme.ask.asset.amount,
+              token: scheme.ask.asset.token,
+              from_account: scheme.participants.buyer,
+              to: asPayee(scheme.participants.seller),
+              then: confirmSwap,
```

Comment:
why do you make the explicit payments instead of relying on how the close works for the account system?